src/app/qgisapp.cpp

```diff
@@ -1988,39 +2001,101 @@ QgsMessageBar *QgisApp::visibleMessageBar()
   }
 }
 
-QList<QgsVectorLayerRef> QgisApp::findBrokenWidgetDependencies( QgsVectorLayer *vl )
+const QList<QgsVectorLayerRef> QgisApp::findBrokenLayerDependencies( QgsVectorLayer *vl, QgsMapLayer::StyleCategories categories ) const
 {
   QList<QgsVectorLayerRef> brokenDependencies;
-  // Check for missing layer widget dependencies
-  for ( int i = 0; i < vl->fields().count(); i++ )
+
+  if ( categories.testFlag( QgsMapLayer::StyleCategory::Forms ) )
+  {
+    // Check for missing layer widget dependencies
+    for ( int i = 0; i < vl->fields().count(); i++ )
+    {
+      std::unique_ptr<QgsEditorWidgetWrapper> ww;
+      ww.reset( QgsGui::editorWidgetRegistry()->create( vl, i, nullptr, nullptr ) );
+      // ww should never be null in real life, but it is in QgisApp tests because
+      // QgsEditorWidgetRegistry widget factories is empty
+      if ( ww )
+      {
+        const QList<QgsVectorLayerRef> constDependencies { ww->layerDependencies() };
+        for ( const QgsVectorLayerRef &dependency : constDependencies )
+        {
+          const QgsVectorLayer *depVl { QgsVectorLayerRef( dependency ).resolveWeakly(
+                                          QgsProject::instance(),
+                                          QgsVectorLayerRef::MatchType::Name ) };
+          if ( ! depVl || ! depVl->isValid() )
+          {
+            brokenDependencies.append( dependency );
+          }
+        }
+      }
+    }
+  }
+
+
+  if ( categories.testFlag( QgsMapLayer::StyleCategory::Relations ) )
   {
-    std::unique_ptr<QgsEditorWidgetWrapper> ww;
-    ww.reset( QgsGui::editorWidgetRegistry()->create( vl, i, nullptr, nullptr ) );
-    // ww should never be null in real life, but it is in QgisApp tests because
-    // QgsEditorWidgetRegistry widget factories is empty
-    if ( ww )
+    // Check for layer weak relations
+    const QList<QgsWeakRelation> constWeakRelations { vl->weakRelations() };
+    for ( const QgsWeakRelation &rel : constWeakRelations )
     {
-      const auto constDependencies { ww->layerDependencies() };
-      for ( const QgsVectorLayerRef &dependency : constDependencies )
+      QgsRelation relation { rel.resolvedRelation( QgsProject::instance(), QgsVectorLayerRef::MatchType::Name ) };
+      QgsVectorLayerRef dependency;
+      bool found = false;
+      if ( ! relation.isValid() )
       {
-        const QgsVectorLayer *depVl { QgsVectorLayerRef( dependency ).resolveWeakly(
-                                        QgsProject::instance(),
-                                        QgsVectorLayerRef::MatchType::Name ) };
-        if ( ! depVl || ! depVl->isValid() )
+        // This is the big question: do we really
+        // want to automatically load the referencing layer(s) too?
```

Comment:
I think yes, we do...

Reply:
Consider this (real) use case: 200 data layers in a PG DB, every data layer uses a small set of referenced aspatial layers (like categories etc.).
The use loads one data layer, the referenced layers is loaded, all other 200 referencing layers are automatically loaded. This is what scares me.